src/emitters/astroobject.cpp

```diff
@@ -0,0 +1,256 @@
+        si.wavelengths          = it.wavelengths;
+
+        UnpolarizedSpectrum weight = depolarizer<Spectrum>(m_irradiance->eval(si, active)) / m_omega / pdf;
+
+        return { ds, weight & active };
+    }
+
+    std::pair<Wavelength, Spectrum>
+    sample_wavelengths(const SurfaceInteraction3f &si, Float sample,
+                       Mask active) const override {
+        return m_irradiance->sample_spectrum(
+            si, math::sample_shifted<Wavelength>(sample), active);
+    }
+
+    std::pair<Ray3f, Spectrum> sample_ray(Float, Float,
+                                          const Point2f &,
+                                          const Point2f &,
+                                          Mask active) const override {
+        MI_MASKED_FUNCTION(ProfilerPhase::EndpointSampleRay, active);
+        if constexpr (dr::is_jit_v<Float>) {
+            /* Do not throw an exception in JIT-compiled variants. This
+               function might be invoked by DrJit's virtual function call
+               recording mechanism despite not influencing any actual
+               calculation. */
+            return { dr::zeros<Ray3f>(), dr::NaN<Spectrum> };
+        } else {
+            NotImplementedError("sample_ray");
+        }
+    }
+
+    Float pdf_direction(const Interaction3f & /*it*/, const DirectionSample3f &ds,
+                        Mask active) const override {
+        MI_MASKED_FUNCTION(ProfilerPhase::EndpointEvaluate, active);
+
+        Vector3f d = m_to_world.value().inverse().transform_affine(ds.d);
+        Float pdf = warp::square_to_uniform_cone_pdf(d, m_angular_radius_cos);
+
+        return pdf;
+    }
+
+    std::pair<PositionSample3f, Float>
+    sample_position(Float /*time*/, const Point2f & /*sample*/,
+                    Mask /*active*/) const override {
+        if constexpr (dr::is_jit_v<Float>) {
+            // When vcalls are recorded in symbolic mode, we can't throw an exception,
+            // even though this result will be unused.
+            return { dr::zeros<PositionSample3f>(),
+                     dr::full<Float>(dr::NaN<Float>) };
+        } else {
+            NotImplementedError("sample_position");
+        }
+    }
+
+    ScalarBoundingBox3f bbox() const override {
+        /* This emitter does not occupy any particular region
+           of space, return an invalid bounding box */
+        return ScalarBoundingBox3f();
+    }
+
+    std::string to_string() const override {
+        std::ostringstream oss;
+        oss << "AstroObjectEmitter[" << std::endl
+            << "  irradiance = " << string::indent(m_irradiance) << ","
+            << std::endl
+            << "  bsphere = " << string::indent(m_bsphere) << "," << std::endl
+            << "  angular_radius_cos = " << string::indent(m_angular_radius_cos) << ","
+            << std::endl
+            << "  omega = " << string::indent(m_omega) << ","
+            << std::endl
+            << "]";
```

Comment:
A more useful repr would IMO report the following:
* `irradiance`
* bounding sphere (consistency with `directional`)
* angular size (`cos_angular_radius` is okay)
* direction

Reporting `omega` is not very useful IMO since it's just a proxy of other data.

Reply:
This was convenient for debugging but I think the debug logs in the constructor should be enough for that